README.md
# net-denon-gem
A Ruby library for controlling Denon AV receivers via their embedded
telnet protocol.

Acknowledgements: Unit tests have been made possible by abstractions and
concepts borrowed from Net::SSH.

## FEATURES/PROBLEMS:

* pre-alpha - use at your own risk
* only tested with the Denon AVR-4308CI

## SYNOPSIS:

  FIXME (code sample of usage)

## REQUIREMENTS:

* FIXME (list of requirements)

## INSTALL:

* sudo gem install net-denon

## LICENSE:

(The MIT License)

Copyright (c) 2008 Jeff Hutchison

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
== 1.0.0 / 2008-06-21

* 1 major enhancement
  * Birthday!


test/common.rb
$LOAD_PATH.unshift "#{File.dirname(__FILE__)}/../lib"
require 'rubygems'
require 'test/unit'
require 'mocha'
require 'net/denon/constants'
require 'net/denon/loggable'

include Net::Denon::Constants

def power_event(state=on)
  (state ? POWER_ON : POWER_STANDBY) + CR
end

def mute_event(state=on)
  (state ? MUTE_ON : MUTE_OFF) + CR
end

def volume_event(vol, max)
  MASTER_VOLUME_SET + vol.to_s + CR + MASTER_VOLUME_MAX + max.to_s + CR
end

def input_source_event(source)
  case source
  when :phono
    INPUT_SOURCE_PHONO + CR
  when :cd
    INPUT_SOURCE_CD + CR
  when :tuner
    INPUT_SOURCE_TUNER + CR
  when :dvd
    INPUT_SOURCE_DVD + CR
  when :hdp
    INPUT_SOURCE_HDP + CR
  when :tv_cbl
    INPUT_SOURCE_TV_CBL + CR
  when :sat
    INPUT_SOURCE_SAT + CR
  when :vcr
    INPUT_SOURCE_VCR + CR
  when :dvr
    INPUT_SOURCE_DVR + CR
  when :v_aux
    INPUT_SOURCE_V_AUX + CR
  when :net_usb
    INPUT_SOURCE_NET_USB + CR
  when :xm
    INPUT_SOURCE_XM + CR
  when :hdradio
    INPUT_SOURCE_HDRADIO + CR
  when :dab
    INPUT_SOURCE_DAB + CR
  when :ipod
    INPUT_SOURCE_IPOD + CR
  else
    raise ArgumentError.new("input source not recognized: #{source}")
  end

end

def volume_command(vol)
  MASTER_VOLUME_SET + vol.to_s
end



test/test_basic.rb
require "test/unit"

require File.expand_path(
    File.join(File.dirname(__FILE__), %w[.. lib net denon]))

require File.join(File.dirname(__FILE__), %w[mock_socket])

class TestBasic < Test::Unit::TestCase

  def test_should_initialize
    proxy = MockSocket.new()
    denon = Net::Denon::Session::new(:proxy => proxy)
    assert_instance_of(Net::Denon::Session, denon)
    denon.close
  end
  
  def test_should_open_debug_log
    proxy = MockSocket.new()
    denon = Net::Denon::start(:proxy => proxy, :log => "debug.txt")
    denon.close
    assert(denon.closed?, "should be closed")
  end
  
  def test_should_not_connect_to_localhost
    # this will fail if you have a telnet server running
    assert_raise(Errno::ECONNREFUSED) do
      denon = Net::Denon::start(:host => "localhost", :log => "debug.txt")
    end
  end
  
  def test_should_timeout_connecting
    assert_raise(Timeout::Error) do
      denon = Net::Denon::start(:host => "www.google.com", :log => "debug.txt")    
    end
  end
end

test/test_state.rb
require 'common'
require 'net/denon/state'
require 'net/denon/constants'

class TestState < Test::Unit::TestCase
  include Net::Denon::Constants
  
  def setup
    @ds = Net::Denon::State.new
  end

  def test_initialize
    assert_nil(@ds.standby?)
    assert_nil(@ds.mute?)
  end
  
  def test_should_update_power
    @ds.update(POWER_STANDBY + CR)
    assert(@ds.standby?)
    assert(!@ds.on?)
    
    @ds.update(POWER_STANDBY + CR + POWER_ON + CR)
    assert(@ds.on?)
    assert(!@ds.standby?)
    
    @ds.update(POWER_ON + CR + POWER_STANDBY + CR)
    assert(!@ds.on?)
    
  end
  
  def test_should_update_master_volume
    @ds.update(MASTER_VOLUME_SET + '00' + CR)
    assert_equal(0, @ds.master_volume)
    @ds.update(MASTER_VOLUME_SET + '10' + CR)
    assert_equal(10, @ds.master_volume)
    @ds.update(MASTER_VOLUME_SET + '00' + CR + MASTER_VOLUME_MAX + '98' + CR)
    assert_equal(98, @ds.master_volume_max)
    assert_equal(00, @ds.master_volume)
    @ds.update(MASTER_VOLUME_SET + '95' + CR + MASTER_VOLUME_MAX + '96' + CR +
      MASTER_VOLUME_SET + '97' + CR + MASTER_VOLUME_MAX + '99' + CR)
    assert_equal(97, @ds.master_volume)
    assert_equal(99, @ds.master_volume_max)
  end
  
  def test_should_update_mute
    @ds.update(MUTE_ON + CR)
    assert(@ds.mute?)
  end
  
  def test_should_update_multiple
    @ds.update(MUTE_OFF + CR + MASTER_VOLUME_SET + '73' + CR)
    assert_equal(73, @ds.master_volume)
    assert(!@ds.mute?)
    @ds.update(MASTER_VOLUME_SET + '21' + CR + MUTE_ON + CR)
    assert_equal(21, @ds.master_volume)
    assert(@ds.mute?)
  end
  
  def test_should_update_channel_volume
    @ds.update("CVFL 23\r")
    assert_equal(23, @ds.channel_volume[:front_left])
    @ds.update("CVFR 24\r")
    assert_equal(24, @ds.channel_volume[:front_right])
    @ds.update("CVC 25\r")
    assert_equal(25, @ds.channel_volume[:center])
    @ds.update("CVSW 26\r")
    assert_equal(26, @ds.channel_volume[:subwoofer])
    @ds.update("CVSL 27\r")
    assert_equal(27, @ds.channel_volume[:surround_left])
    @ds.update("CVSR 28\r")
    assert_equal(28, @ds.channel_volume[:surround_right])
    @ds.update("CVSBL 29\r")
    assert_equal(29, @ds.channel_volume[:surround_back_left])
    @ds.update("CVSBR 30\r")
    assert_equal(30, @ds.channel_volume[:surround_back_right])
    @ds.update("CVSB 31\r")
    assert_equal(31, @ds.channel_volume[:surround_back])
    @ds.update("CVFL 32\rCVFR 33\rCVC 34\r")
    assert_equal(32, @ds.channel_volume[:front_left])
    assert_equal(33, @ds.channel_volume[:front_right])
    assert_equal(34, @ds.channel_volume[:center])
  
  end

  def test_should_update_input_source
    @ds.update(INPUT_SOURCE_PHONO + CR)
    assert_equal("PHONO", @ds.input_source)
    @ds.update(INPUT_SOURCE_NET_USB + CR)
    assert_equal("NET/USB", @ds.input_source)
  end
  
  def test_should_update_main_zone
    @ds.update("ZMON\r")
    assert(@ds.main_zone_on?)
    @ds.update("ZMON\rZMOFF\r")
    assert(!@ds.main_zone_on?)
    
  end
  
  def test_should_update_record_source
    @ds.update("SRPHONO\r")
    assert_equal("PHONO", @ds.record_source)
    @ds.update("SRNET/USB\r")
    assert_equal("NET/USB", @ds.record_source)
  end
  
end

lib/net-denon-gem.rb
# $Id$

# Equivalent to a header guard in C/C++
# Used to prevent the class/module from being loaded more than once
unless defined? NetDenonGem

module NetDenonGem

  # :stopdoc:
  VERSION = '1.0.0'
  LIBPATH = ::File.expand_path(::File.dirname(__FILE__)) + ::File::SEPARATOR
  PATH = ::File.dirname(LIBPATH) + ::File::SEPARATOR
  # :startdoc:

  # Returns the version string for the library.
  #
  def self.version
    VERSION
  end

  # Returns the library path for the module. If any arguments are given,
  # they will be joined to the end of the libray path using
  # <tt>File.join</tt>.
  #
  def self.libpath( *args )
    args.empty? ? LIBPATH : ::File.join(LIBPATH, *args)
  end

  # Returns the lpath for the module. If any arguments are given,
  # they will be joined to the end of the path using
  # <tt>File.join</tt>.
  #
  def self.path( *args )
    args.empty? ? PATH : ::File.join(PATH, *args)
  end

  # Utility method used to rquire all files ending in .rb that lie in the
  # directory below this file that has the same name as the filename passed
  # in. Optionally, a specific _directory_ name can be passed in such that
  # the _filename_ does not have to be equivalent to the directory.
  #
  def self.require_all_libs_relative_to( fname, dir = nil )
    dir ||= ::File.basename(fname, '.*')
    search_me = ::File.expand_path(
        ::File.join(::File.dirname(fname), dir, '**', '*.rb'))

    Dir.glob(search_me).sort.each {|rb| require rb}
  end

end  # module NetDenonGem

NetDenonGem.require_all_libs_relative_to __FILE__

end  # unless defined?

# EOF


lib/net/denon/state.rb
require 'net/denon/constants'
require 'net/denon/loggable'

module Net ; module Denon

  class State
    include Constants
    include Loggable

    attr_reader :master_volume
    
    attr_reader :master_volume_max
    
    attr_reader :input_source
    
    attr_reader :channel_volume
    
    attr_reader :record_source
  
    def initialize
      @channel_volume = Hash.new
      @standby        = nil
      @mute           = nil
    end

    def standby?
      debug {"@standby = #{@standby}"}
      @standby
    end

    def on?
      ! @standby
    end

    def mute?
      @mute
    end
  
    def main_zone_on?
      @main_zone
    end

    def update(response)
      debug {"response = #{response}"}
      response.each("\r") do |r|
        case command(r)
        when POWER
          update_power(r)
        when MASTER_VOLUME
          update_master_volume(r)
        when CHANNEL_VOLUME
          update_channel_volume(r)
        when MUTE
          update_mute(r)
        when INPUT_SOURCE
          update_input_source(r)
        when MAIN_ZONE
          update_main_zone(r)
        when RECORD_SOURCE
          update_record_source(r)
        end
      end
    end

    private

    def update_power(r)
      @standby = (parameter(r) == "STANDBY")
    end

    def update_master_volume(r)
      p = parameter(r)
      if p.length == 2 then
        @master_volume = p.to_i
      else
        @master_volume_max = p[-2..-1].to_i
      end
    end

    def update_channel_volume(r)
      re = /^([A-Z]+) (\d\d)(\d?)$/
      md = re.match(parameter(r))
      vol = md[2].to_i
      case (md[1])
      when "FL"
        @channel_volume[:front_left] = vol
      when "FR"
        @channel_volume[:front_right] = vol
      when "C"
        @channel_volume[:center] = vol
      when "SW"
        @channel_volume[:subwoofer] = vol
      when "SL"
        @channel_volume[:surround_left] = vol
      when "SR"
        @channel_volume[:surround_right] = vol
      when "SBL"
        @channel_volume[:surround_back_left] = vol
      when "SBR"
        @channel_volume[:surround_back_right] = vol
      when "SB"
        @channel_volume[:surround_back] = vol
      end
    end

    def update_mute(r)
      @mute = (parameter(r) == "ON")
    end

    def update_input_source(r)
        @input_source = parameter(r)
    end
  
    def update_main_zone(r)
      @main_zone = (parameter(r) == "ON")
    end
  
    def update_record_source(r)
        @record_source = parameter(r)
    end
  
    def command(r)
      r[0, 2]
    end

    def parameter(r)
      r[2, r.length-3]
    end

    def to_s
     string = ''
     string += "standby: #{@standby}\n" 
     string += "mute: #{@mute}\n"
     string += "master volume: #{@master_volume}\n"
     string += "master max volume: #{@master_max_volume}\n"
     string += "input source: #{@input_source}\n"
     string += "main zone: #{@main_zone}\n"
     string += "record source: #{record_source}\n"
    end
  end
end ; end

lib/net/denon/status.rb
require 'net/denon/constants'

module Net ; module Denon

  class Status
    include Constants

    attr_reader :master_volume
    
    attr_reader :master_volume_max
    
    attr_reader :input_source
    
    attr_reader :channel_volume
    
    attr_reader :record_source
  
    def initialize
      @channel_volume = Hash.new
      @standby        = nil
      @mute           = nil
    end

    def standby?
      @standby
    end

    def on?
      ! @standby
    end

    def mute?
      @mute
    end
  
    def main_zone_on?
      @main_zone
    end

    def update(response)
      response.each("\r") do |r|
        case command(r)
        when POWER
          update_power(r)
        when MASTER_VOLUME
          update_master_volume(r)
        when CHANNEL_VOLUME
          update_channel_volume(r)
        when MUTE
          update_mute(r)
        when INPUT_SOURCE
          update_input_source(r)
        when MAIN_ZONE
          update_main_zone(r)
        when RECORD_SOURCE
          update_record_source(r)
        end
      end
    end

    private

    def update_power(r)
      @standby = (parameter(r) == "STANDBY")
    end

    def update_master_volume(r)
      p = parameter(r)
      if p.length == 2 then
        @master_volume = p.to_i
      else
        @master_volume_max = p[-2..-1].to_i
      end
    end

    def update_channel_volume(r)
      re = /^([A-Z]+) (\d\d)(\d?)$/
      md = re.match(parameter(r))
      vol = md[2].to_i
      case (md[1])
      when "FL"
        @channel_volume[:front_left] = vol
      when "FR"
        @channel_volume[:front_right] = vol
      when "C"
        @channel_volume[:center] = vol
      when "SW"
        @channel_volume[:subwoofer] = vol
      when "SL"
        @channel_volume[:surround_left] = vol
      when "SR"
        @channel_volume[:surround_right] = vol
      when "SBL"
        @channel_volume[:surround_back_left] = vol
      when "SBR"
        @channel_volume[:surround_back_right] = vol
      when "SB"
        @channel_volume[:surround_back] = vol
      end
    end

    def update_mute(r)
      @mute = (parameter(r) == "ON")
    end

    def update_input_source(r)
        @input_source = parameter(r)
    end
  
    def update_main_zone(r)
      @main_zone = (parameter(r) == "ON")
    end
  
    def update_record_source(r)
        @record_source = parameter(r)
    end
  
    def command(r)
      r[0, 2]
    end

    def parameter(r)
      r[2, r.length-3]
    end

    def to_s
     string = ''
     string += "standby: #{@standby}\n" 
     string += "mute: #{@mute}\n"
     string += "master volume: #{@master_volume}\n"
     string += "master max volume: #{@master_max_volume}\n"
     string += "input source: #{@input_source}\n"
     string += "main zone: #{@main_zone}\n"
     string += "record source: #{record_source}\n"
    end
  end
end ; end

lib/net/denon/loggable.rb
module Net; module Denon

  # A simple module to make logging easier to deal with. It assumes that the
  # logger instance (if not nil) quacks like a Logger object (in Ruby's
  # standard library). Borrowed with gratitude from Net::SSH.
  #
  #   class MyClass
  #     include Net::Denon::Loggable
  #   end
  #
  module Loggable
    # The logger instance that will be used to log messages. If nil, nothing
    # will be logged.
    attr_accessor :logger

    # Displays the result of yielding if the log level is Logger::DEBUG or
    # greater.
    def debug
      logger.add(Logger::DEBUG, nil, facility) { yield } if logger
    end

    # Displays the result of yielding if the log level is Logger::INFO or
    # greater.
    def info
      logger.add(Logger::INFO, nil, facility) { yield } if logger
    end

    # Displays the result of yielding if the log level is Logger::WARN or
    # greater. (Called lwarn to avoid shadowing with Kernel#warn.)
    def lwarn
      logger.add(Logger::WARN, nil, facility) { yield } if logger
    end

    # Displays the result of yielding if the log level is Logger:ERROR or
    # greater.
    def error
      logger.add(Logger::ERROR, nil, facility) { yield } if logger
    end

    # Displays the result of yielding if the log level is Logger::FATAL or
    # greater.
    def fatal
      logger.add(Logger::FATAL, nil, facility) { yield } if logger
    end

    private

      # Sets the "facility" value, used for reporting where a log message
      # originates. It defaults to the name of class with the object_id
      # appended.
      def facility
        @facility ||= self.class.name.gsub(/::/, ".").gsub(/([a-z])([A-Z])/, "\\1_\\2").downcase + "[%x]" % object_id
      end
  end
end; end

lib/net/denon/constants.rb
module Net; module Denon

  # Definitions of constants.
  module Constants
    
    CR                   = "\r"

    POWER                = 'PW'
    POWER_ON             = 'PWON'
    POWER_STANDBY        = 'PWSTANDBY'
    POWER_STATUS         = 'PW?'

    MASTER_VOLUME        = 'MV'
    MASTER_VOLUME_UP     = 'MVUP'
    MASTER_VOLUME_DOWN   = 'MVDOWN'
    MASTER_VOLUME_SET    = 'MV'
    MASTER_VOLUME_STATUS = 'MV?'
    MASTER_VOLUME_MAX    = 'MVMAX'
    
    CHANNEL_VOLUME       = 'CV'

    MUTE                 = 'MU'
    MUTE_ON              = 'MUON'
    MUTE_OFF             = 'MUOFF'
    MUTE_STATUS          = 'MU?'

    INPUT_SOURCE         = 'SI'
    INPUT_SOURCE_PHONO   = 'SIPHONO'
    INPUT_SOURCE_CD      = 'SICD'
    INPUT_SOURCE_TUNER   = 'SITUNER'
    INPUT_SOURCE_DVD     = 'SIDVD'
    INPUT_SOURCE_HDP     = 'SIHDP'
    INPUT_SOURCE_TV_CBL  = 'SITV/CBL'
    INPUT_SOURCE_SAT     = 'SISAT'
    INPUT_SOURCE_VCR     = 'SIVCR'
    INPUT_SOURCE_DVR     = 'SIDVR'
    INPUT_SOURCE_V_AUX   = 'SIV.AUX'
    INPUT_SOURCE_NET_USB = 'SINET/USB'
    INPUT_SOURCE_XM      = 'SIXM'
    INPUT_SOURCE_HDRADIO = 'SIHDRADIO'
    INPUT_SOURCE_DAB     = 'SIDAB'
    INPUT_SOURCE_IPOD    = 'SIIPOD'
    INPUT_SOURCE_STATUS  = 'SI?'

    MAIN_ZONE            = 'ZM'

    RECORD_SOURCE        = 'SR'

  end
  
end; end

lib/net/denon/transport.rb
require 'socket'
require 'timeout'
require 'net/denon/loggable'

module Net; module Denon
  
  class Transport
    include Loggable
    
    # The standard telnet port, will be used unless specified in options
    DEFAULT_PORT = 23
    
    # The IP address that was passed to constructor.
    attr_reader :host
    
    # The port connected to. Defaults to DEFAULT_PORT.
    attr_reader :port
    
    # The underlying socket object being used to communicate with the
    # receiver.
    attr_reader :socket
    
    # The options passed to constructor.
    attr_reader :options

    # Instantiates a new Transport object used to communicate with the
    # receiver. Options include:
    #  port:: the port to connect to, defaults to DEFAULT_PORT.
    #  logger:: the logger to use
    #  proxy:: a proxy class for TCPSocket
    #  timeout:: amount of time to wait for connection, defaults to no timeout
    def initialize(host, options={})
      self.logger = options[:logger]
      @host       = host
      @port       = options[:port] || DEFAULT_PORT
      @options    = options
      
      debug { "establishing connection to #{@host}:#{@port}" }
      factory = options[:proxy] || TCPSocket
      @socket = Timeout::timeout(options[:timeout] || 0) { factory.open(@host, @port) }

      debug { "connection established" }
    end
    
    # Closes the connection to the receiver.
    def close
      socket.close
    end

    # Sends a command string to the receiver.
    def send(string)
      info { "sending: #{string}" }
      string += "\r"
      length = string.length
      while 0 < length
        IO::select(nil, [socket])
        length -= socket.syswrite(string[-length..-1])
      end
    end
    
    # Tries to read the next event(s) from the socket.
    def poll_events
      buffer = ''
      line = "\r"

      until(line[-1] == 13 and not IO::select([socket], nil, nil, @options[:wait_time] || 0.1))
        buffer = socket.readpartial(1024)
        line += buffer
      end
      info { "received: #{line}" }
      line
    end
      
  end
end ; end

lib/net/denon/session.rb
require 'socket'
require 'timeout'
require 'net/denon/state'
require 'net/denon/transport'
require 'net/denon/constants'
require 'net/denon/loggable'

module Net ; module Denon

  #
  # == Net::Denon::Session
  #
  # Provides DENON AVR control protocol client functionality.
  #
  # == Overview
  #
  # The DENON AVR control protocol allows a client to send commands to a Denon
  # AVR reciever via its embedded telnet server. This library currently
  # implements version 5.1.0_a for model AVR-4308.
  #
  # == Examples
  #
  # === Log in and send a command
  #
  #   denon = Net::Denon.new("10.0.1.201")
  #   denon.power_on
  #   denon.close
  #
  # == References
  #
  # This library implements the DENON AVR control protocol as documented by
  # Denon at:
  # http://usa.denon.com/AVR-4308CISerialProtocol_Ver5.1.0a.pdf
  #
  class Session
    include Constants
    include Loggable
    
    # The underlying transport.
    attr_reader :transport
    
    # The current state of the receiver
    attr_reader :state
  
    # Creates a new Net::Denon object and connects it to the telnet port (23) of
    # the Denon receiver on the named host.
    #
    # +options+ is a hash of options.  The following example lists all options
    # and their default values.
    #
    #   denon = Net::Denon::new(
    #     :host       => "10.0.1.201", # default: nil
    #     :port       => 23,           # default: 23
    #     :wait_time  => 0.2,          # default: 0.2 sec
    #     :timeout    => 1,            # default: 1 sec
    #     :log        => "debug.txt",  # default: nil (no output)
    #     :proxy      => proxy,        # default: nil
    #     )
    #
    # The options have the following meanings:
    #
    # host:: the hostname or IP address to connect to, as a String.
    #
    # port:: the port to connect to; defaults to 23.
    #
    # wait_time:: the amount of time to wait for a response after sending
    #             a command; default is 0.2 sec.
    #
    # timeout:: the amount of time to wait trying to connect.  Exceeding
    #           this timeout value causes a Timeout::Error to be raised.
    #           Defaults to 1 sec.
    #
    # log:: commands and output will be dumped to this file; defaults
    #       to no output.
    #
    # proxy:: a proxy object to be used instead of opening a direct connection
    #         to the host.
    #
    def initialize(host, options)
      self.logger = options[:logger]
      @transport = Net::Denon::Transport.new(host, options)
      @state = Net::Denon::State.new
    end
  
    # Disconnects from the server.
    def close
      transport.close
    end
  
    def query!
      send_command POWER_STATUS
      send_command MUTE_STATUS
      send_command MASTER_VOLUME_STATUS
      send_command INPUT_SOURCE_STATUS
    end
    
    def on!
      send_command POWER_STATUS
      send_command POWER_ON unless state.on?
    end
    
    def standby!
      send_command POWER_STATUS
      send_command POWER_STANDBY unless state.standby?
    end
    
    def mute!
      send_command MUTE_STATUS
      send_command MUTE_ON unless state.mute?
    end
    
    def unmute!
      send_command MUTE_STATUS
      send_command MUTE_OFF if state.mute?
    end
    
    def master_volume=(volume)
      v = volume.to_i
      if ((0..99) === v) then
        send_command(MASTER_VOLUME_SET + v.to_s)
      else
        raise ArgumentError.new("can't set volume to #{volume}")
      end
    end

    def input_source(source)
      case source
      when :phono
        send_command INPUT_SOURCE_PHONO
      when :cd
        send_command INPUT_SOURCE_CD
      when :tuner
        send_command INPUT_SOURCE_TUNER
      when :dvd
        send_command INPUT_SOURCE_DVD
      when :hdp
        send_command INPUT_SOURCE_HDP
      when :tv_cbl
        send_command INPUT_SOURCE_TV_CBL
      when :sat
        send_command INPUT_SOURCE_SAT
      when :vcr
        send_command INPUT_SOURCE_VCR
      when :dvr
        send_command INPUT_SOURCE_DVR
      when :v_aux
        send_command INPUT_SOURCE_V_AUX
      when :net_usb
        send_command INPUT_SOURCE_NET_USB
      when :xm
        send_command INPUT_SOURCE_XM
      when :hdradio
        send_command INPUT_SOURCE_HDRADIO
      when :dab
        send_command INPUT_SOURCE_DAB
      when :ipod
        send_command INPUT_SOURCE_IPOD
      else
        raise ArgumentError.new("input source not recognized: #{source}")
      end
    end
    
    def send_command(command)
      debug { "send_command called: #{command}" }
      transport.send command
      sleep 0.1
      update_state
    end

    protected
    
    def update_state
      debug { "check_status called" }
      state.update(transport.poll_events)
    end
  
  end

end ; end